src/librustdoc/html/static/js/search.js

```diff
@@ -167,33 +484,138 @@ window.initSearch = function(rawSearchIndex) {
-     *   returned: Array<?>,
-     *   others: Array<?>
-     * }}
+     * Parses the query.
+     *
+     * The supported syntax by this parser is as follow:
+     *
+     * ident = *(ALPHA / DIGIT / "_")
+     * path = ident *(DOUBLE-COLON ident)
+     * arg = path [generics]
+     * arg-without-generic = path
+     * nonempty-arg-list = arg *WS *(COMMA/WS *WS arg)
+     * nonempty-arg-list-without-generics = arg-without-generic *WS *(COMMA/WS *WS arg
+     *                                                                without-generic)
+     * generics = OPEN-ANGLE-BRACKET *WS [ nonempty-arg-list-without-generics ] *WS
+     *            CLOSE-ANGLE-BRACKET
+     * return-args = RETURN-ARROW *WS nonempty-arg-list
+     *
+     * exact-search = [type-filter *WS COLON] *WS QUOTE ident QUOTE *WS [generics]
+     * type-search = [type-filter *WS COLON] *WS path *WS generics
+     *
+     * query = *WS (exact-search / type-search / return-args) *WS
+     *
+     * type-filter = (
+     *     "mod" /
+     *     "externcrate" /
+     *     "import" /
+     *     "struct" /
+     *     "enum" /
+     *     "fn" /
+     *     "type" /
+     *     "static" /
+     *     "trait" /
+     *     "impl" /
+     *     "tymethod" /
+     *     "method" /
+     *     "structfield" /
+     *     "variant" /
+     *     "macro" /
+     *     "primitive" /
+     *     "associatedtype" /
+     *     "constant" /
+     *     "associatedconstant" /
+     *     "union" /
+     *     "foreigntype" /
+     *     "keyword" /
+     *     "existential" /
+     *     "attr" /
+     *     "derive" /
+     *     "traitalias")
+     *
+     * OPEN-ANGLE-BRACKET = "<"
+     * CLOSE-ANGLE-BRACKET = ">"
+     * COLON = ":"
+     * DOUBLE-COLON = "::"
+     * QUOTE = %x22
+     * COMMA = ","
+     * RETURN-ARROW = "->"
+     *
+     * ALPHA = %x41-5A / %x61-7A ; A-Z / a-z
+     * DIGIT = %x30-39
+     * WS = %x09 / " "
+     *
+     * @param  {string} val     - The user query
+     *
+     * @return {ParsedQuery}    - The parsed query
      */
-    function execQuery(query, searchWords, filterCrates) {
-        function itemTypeFromName(typename) {
-            for (var i = 0, len = itemTypes.length; i < len; ++i) {
-                if (itemTypes[i] === typename) {
-                    return i;
-                }
+    function parseQuery(userQuery) {
+        userQuery = userQuery.trim();
+        var parserState = {
+            length: userQuery.length,
+            pos: 0,
+            // Total number of elements (includes generics).
+            totalElems: 0,
+            typeFilter: null,
+            userQuery: userQuery.toLowerCase(),
+        };
+        var query = newParsedQuery(userQuery);
+
+        try {
+            parseInput(query, parserState);
+            if (parserState.typeFilter !== null) {
+                var typeFilter = parserState.typeFilter.replace(/^const$/, "constant");
+                query.typeFilter = itemTypeFromName(typeFilter);
```

Comment:
This seems like an unfortunate way of implementing an alias of `const` to `constant`.

First of all, it's part of the syntax, but is not reflected in the eBNF.

Second, using a regex for this seems like overkill.

```suggestion
                var typeFilter = parserState.typeFilter;
                if (typeFilter === "const") { typeFilter = "constant"; }
                query.typeFilter = itemTypeFromName(typeFilter);
```

Reply:
Good point. I simply moved this code here but this improvement will be very welcome.